### PR TITLE
WaterRenderObjClass

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/client/water/w3dwater.cpp [hunk 1]
```diff
@@ -0,0 +1,2327 @@
...
+    m_sortLevel = level;
+}
+
+void WaterRenderObjClass::Xfer_Snapshot(Xfer *xfer)
+{
+    uint8_t version = 1;
+    xfer->xferVersion(&version, 1);
+    int x = m_gridCellsX;
+    xfer->xferInt(&x);
+    captainslog_assert(x != m_gridCellsX);
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: shouldn't these be other way around ==? they should also state what's wrong, use captainslog_dbgassert for asserts with messages

File: src/game/client/water/w3dwater.cpp [hunk 2]
```diff
@@ -0,0 +1,2327 @@
...
+    poly.Add_Vertex(Vector3(m_dx, 0.0f, m_level));
+    poly.Clip(f.m_planes[0], poly2);
+    poly2.Clip(f.m_planes[1], poly);
+    poly.Clip(f.m_planes[2], poly2);
+    poly2.Clip(f.m_planes[3], poly);
+    poly.Clip(f.m_planes[4], poly2);
+    poly2.Clip(f.m_planes[5], poly);
+    int count = poly.m_verts.Count();
+
+    if (count < 3) {
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: flip it

File: src/game/client/water/w3dwater.cpp [hunk 3]
```diff
@@ -0,0 +1,2327 @@
...
+            }
+        }
+    }
+#endif
+}
+
+bool WaterRenderObjClass::Get_Clipped_Water_Plane(CameraClass *cam, AABoxClass *box)
+{
+    const FrustumClass &f = cam->Get_Frustum();
+    ClipPolyClass poly;
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: poly1 for consistency

File: src/game/client/water/w3dwater.cpp [hunk 4]
```diff
@@ -0,0 +1,2327 @@
...
+
+    m_meshVertexMaterialClass = new VertexMaterialClass();
+    m_meshVertexMaterialClass->Set_Shininess(20.0f);
+    m_meshVertexMaterialClass->Set_Ambient(1.0f, 1.0f, 1.0f);
+    m_meshVertexMaterialClass->Set_Diffuse(1.0f, 1.0f, 1.0f);
+    m_meshVertexMaterialClass->Set_Specular(0.5f, 0.5f, 0.5f);
+    m_meshVertexMaterialClass->Set_Opacity(0.5f);
+    m_meshVertexMaterialClass->Set_Lighting(true);
+
+    Load_Setting(&m_settings[1], TIME_OF_DAY_MORNING);
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: use TOD enum for the numbers

File: src/game/client/water/w3dwater.h [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    unsigned long m_riverWaterPixelShader;
+    unsigned long m_trapezoidWaterPixelShader;
+    TextureClass *m_waterSparklesTexture;
+    float m_riverXOffset;
+    float m_riverYOffset;
+    bool m_drawingRiver;
+    bool m_disableRiver;
+    TextureClass *m_riverAlphaEdge;
+    TimeOfDayType m_tod;
+    Setting m_settings[5];
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.h]: 5 is TOD_COUNT

File: src/game/client/water/w3dwater.cpp [hunk 6]
```diff
@@ -0,0 +1,2327 @@
...
+    setting->water_texture = W3DAssetManager::Get_Instance()->Get_Texture(g_waterSettings[tod].m_waterTextureFile);
+    setting->sky_texels_per_unit = g_waterSettings[tod].m_skyTexelsPerUnit;
+    SurfaceClass::SurfaceDescription surface_desc;
+    setting->water_texture->Get_Level_Description(surface_desc, 0);
+    setting->sky_texels_per_unit = setting->sky_texels_per_unit / surface_desc.width;
+    setting->water_repeat_count = g_waterSettings[tod].m_waterRepeatCount;
+    setting->u_scroll_per_ms = g_waterSettings[tod].m_uScrollPerMs;
+    setting->v_scroll_per_ms = g_waterSettings[tod].m_vScrollPerMs;
+
+    setting->vertex00_diffuse = g_waterSettings[tod].m_vertex00Diffuse.blue
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp:857]: Add a TODO to change to use a packer function

File: src/game/client/water/w3dwater.cpp [hunk 7]
```diff
@@ -0,0 +1,2327 @@
...
+
+    if (g_theGameLogic) {
+        frame = g_theGameLogic->Get_Frame();
+    }
+
+    m_riverVOrigin = m_riverVOrigin + 0.0020000001f;
+    m_riverXOffset = m_riverXOffset + 0.000082500002f;
+    m_riverYOffset = m_riverYOffset + 0.000165f;
+
+    if (m_riverXOffset > 1.0) {
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp:750]: -= and += for these no?

File: src/game/client/water/w3dwater.cpp [hunk 8]
```diff
@@ -0,0 +1,2327 @@
...
+
+                            if (m->preferred_height <= m->height) {
+                                m->velocity = g_theWriteableGlobalData->m_gravity * 3.0f + m->velocity;
+                            } else {
+                                m->velocity = m->velocity - g_theWriteableGlobalData->m_gravity * 3.0f;
+                            }
+
+                            m->height = m->height + m->velocity;
+
+                            if (GameMath::Fabs(m->height - m->preferred_height) >= 1.0f
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: flip

File: src/game/client/water/w3dwater.cpp [hunk 9]
```diff
@@ -0,0 +1,2327 @@
...
+void WaterRenderObjClass::Update()
+{
+    unsigned int frame = 0;
+    static unsigned int lastLogicFrame;
+
+    if (g_theGameLogic) {
+        frame = g_theGameLogic->Get_Frame();
+    }
+
+    m_riverVOrigin = m_riverVOrigin + 0.0020000001f;
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: hhmmmm
https://www.wolframalpha.com/input/?i=0.0020000001

33 probably isn't a coincidence here
https://www.wolframalpha.com/input/?i=0.000082500002
https://www.wolframalpha.com/input/?i=0.000165
edit:
yeah its not GameClient::init sets 33.333332 as Frame Rate and TheW3DFrameLengthInMsec defaults to 33

File: src/game/client/water/w3dwater.cpp [hunk 10]
```diff
@@ -0,0 +1,2327 @@
...
+                int y = m_gridCellsY + 1;
+                m_meshInMotion = false;
+                WaterMeshData *m = m_meshData;
+
+                for (int i = 0; i < y + 1; i++) {
+                    for (int j = 0; j < x + 1; j++) {
+                        if ((m->status & 1) != 0) {
+                            m->velocity = 0.93000001f * m->velocity;
+
+                            if (m->preferred_height <= m->height) {
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: flip

File: src/game/client/water/w3dwater.cpp [hunk 11]
```diff
@@ -0,0 +1,2327 @@
...
+
+    if (lastLogicFrame != frame) {
+        if (m_doWaterGrid) {
+            if (m_meshInMotion) {
+                int x = m_gridCellsX + 1;
+                int y = m_gridCellsY + 1;
+                m_meshInMotion = false;
+                WaterMeshData *m = m_meshData;
+
+                for (int i = 0; i < y + 1; i++) {
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: these add 2 not 1
not sure what's the logics there, seems like a bug?

File: src/game/client/water/w3dwater.cpp [hunk 12]
```diff
@@ -0,0 +1,2327 @@
...
+    m_useCloudLayer(true),
+    m_waterType(WATER_TYPE_0_TRANSLUCENT),
+    m_tod(TIME_OF_DAY_AFTERNOON),
+    m_pReflectionTexture(nullptr),
+    m_skyBox(nullptr),
+    m_vertexBufferD3D(nullptr),
+    m_indexBufferD3D(nullptr),
+    m_vertexBufferD3DOffset(0),
+    m_dwWavePixelShader(0),
+    m_meshData(nullptr),
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp:74]: missing m_dwWaveVertexShader

File: src/game/client/water/w3dwater.cpp [hunk 13]
```diff
@@ -0,0 +1,2327 @@
...
+    m_meshInMotion(false),
+    m_gridOrigin(0.0f, 0.0f),
+    m_gridDirectionX(1.0f, 0.0f),
+    m_gridDirectionY(1.0f, 0.0f),
+    m_gridCellSize(10.0f),
+    m_gridCellsX(128),
+    m_gridCellsY(128),
+    m_gridWidth(m_gridCellsX * m_gridCellSize),
+    m_gridHeight(m_gridCellsY * m_gridCellSize),
+    m_riverVOrigin(0.0f),
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp:85]: missing m_pBumpTexture m_pBumpTexture2 or compiler makes it?

File: src/game/client/water/w3dwater.cpp [hunk 14]
```diff
@@ -0,0 +1,2327 @@
...
+    DX8Wrapper::Draw_Triangles(0, 2, 0, 4);
+#endif
+}
+
+void WaterRenderObjClass::Render_Sky_Body(Matrix3D *mat)
+{
+#ifdef BUILD_WITH_D3D8
+    Vector3 v1;
+    Vector3 v2;
+    Vector3 v3;
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: 3 and 4 aren't used?

File: src/game/client/water/w3dwater.cpp [hunk 15]
```diff
@@ -0,0 +1,2327 @@
...
+    m.Set_Translation(Vector3(0.0f, 0.0f, 0.0f));
+    DX8Wrapper::Set_Transform(D3DTS_WORLD, m);
+    DX8Wrapper::Draw_Triangles(0, 2, 0, 4);
+#endif
+}
+
+void WaterRenderObjClass::Render_Sky_Body(Matrix3D *mat)
+{
+#ifdef BUILD_WITH_D3D8
+    Vector3 v1;
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: rename to translation

File: src/game/client/water/w3dwater.cpp [hunk 16]
```diff
@@ -0,0 +1,2327 @@
...
+    Vector3 v2;
+    Vector3 v3;
+    Vector3 v4;
+    Vector3 v5(150.0f, 550.0f, 30.0f);
+    mat->Get_Translation(&v1);
+    v2 = v1 - v5;
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: v2 = v1 - Vector3(150.0f, 550.0f, 30.0f);

File: src/game/client/water/w3dwater.cpp [hunk 17]
```diff
@@ -0,0 +1,2327 @@
...
+            if (i->Point_In_Trigger(c) && i->Get_Point(0)->z >= f) {
+                f = i->Get_Point(0)->z;
+                h = i->Get_Water_Handle();
+            }
+        }
+    }
+
+    if (h) {
+        return h->m_polygon->Get_Point(0)->z;
+    } else {
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: no need for else

File: src/game/client/water/w3dwater.cpp [hunk 18]
```diff
@@ -0,0 +1,2327 @@
...
+    if (cell_size) {
+        *cell_size = m_gridCellSize;
+    }
+}
+
+float WaterRenderObjClass::Get_Water_Height(float x, float y)
+{
+    const WaterHandle *h = nullptr;
+    float f = 0.0f;
+    ICoord3D c;
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: point or coord

File: src/game/client/water/w3dwater.cpp [hunk 19]
```diff
@@ -0,0 +1,2327 @@
...
+
+    if (cell_size) {
+        *cell_size = m_gridCellSize;
+    }
+}
+
+float WaterRenderObjClass::Get_Water_Height(float x, float y)
+{
+    const WaterHandle *h = nullptr;
+    float f = 0.0f;
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: z or height

File: src/game/client/water/w3dwater.cpp [hunk 20]
```diff
@@ -0,0 +1,2327 @@
...
+    }
+    return 0;
+}
+#endif
+
+#ifdef BUILD_WITH_D3D8
+HRESULT WaterRenderObjClass::Generate_Index_Buffer(int size_x, int size_y)
+{
+    m_numIndices = (2 * size_x + 2) * (size_y - 1) - 2;
+    HRESULT error = m_pDev->CreateIndexBuffer(
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: res

File: src/game/client/water/w3dwater.cpp [hunk 21]
```diff
@@ -0,0 +1,2327 @@
...
+        for (int i = 0; i < matinfo->Texture_Count(); i++) {
+            if (matinfo->Peek_Texture(i)) {
+                matinfo->Peek_Texture(i)->Get_Texture_Filter()->Set_U_Address_Mode(
+                    TextureFilterClass::TEXTURE_ADDRESS_CLAMP);
+                matinfo->Peek_Texture(i)->Get_Texture_Filter()->Set_V_Address_Mode(
+                    TextureFilterClass::TEXTURE_ADDRESS_CLAMP);
+            }
+        }
+
+        Ref_Ptr_Release(matinfo);
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: where did this come from? if its a bugfix should be labeled as such

File: src/game/client/water/w3dwater.cpp [hunk 22]
```diff
@@ -0,0 +1,2327 @@
...
+    DWORD fvf = DX8_FVF_XYZDUV2;
+
+    if (do_static) {
+        pool = D3DPOOL_MANAGED;
+        usage = D3DUSAGE_WRITEONLY;
+        fvf = 0;
+    }
+
+    if (!m_vertexBufferD3D) {
+        HRESULT error = m_pDev->CreateVertexBuffer(vertex_size * m_numVertices, usage, fvf, pool, &m_vertexBufferD3D);
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: res

File: src/game/client/water/w3dwater.cpp [hunk 23]
```diff
@@ -0,0 +1,2327 @@
...
+        Generate_Vertex_Buffer(15, 15, 24, true);
+    }
+#endif
+}
+
+void WaterRenderObjClass::Load_Setting(Setting *setting, TimeOfDayType tod)
+{
+    captainslog_assert(setting);
+
+    setting->sky_texture = W3DAssetManager::Get_Instance()->Get_Texture(g_waterSettings[tod].m_skyTextureFile);
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp:846]: have Get_Texture args been confirmed to match defargs?

File: src/game/client/water/w3dwater.cpp [hunk 24]
```diff
@@ -0,0 +1,2327 @@
...
+
+            float maxy = GameMath::Ceil(gy + m_gridChangeMaxRange);
+
+            if (m_gridCellsY < maxy) {
+                maxy = m_gridCellsY;
+            }
+
+            for (int i = miny; i < maxy; i++) {
+                for (int j = minx; j < maxx; j++) {
+                    WaterMeshData *mesh = &m_meshData[(m_gridCellsX + 3) * (i + 1) + 1 + j];
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: 1 + j to j + 1

File: src/game/client/water/w3dwater.cpp [hunk 25]
```diff
@@ -0,0 +1,2327 @@
...
+                    mesh->preferred_height = preferred_height;
+                    mesh->velocity = z_velocity + mesh->velocity;
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: +=

File: src/game/client/water/w3dwater.cpp [hunk 26]
```diff
@@ -0,0 +1,2327 @@
...
+                        ICoord3D p1 = *i->Get_Point(0);
+                        ICoord3D p2 = *i->Get_Point(j);
+                        ICoord3D p3 = *i->Get_Point(j + 1);
+                        ICoord3D p4 = *i->Get_Point(j + 1);
+
+                        if (j + 2 < i->Get_Num_Points()) {
+                            ICoord3D p4 = *i->Get_Point(j + 2);
+                        }
+
+                        Vector3 v[4];
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: points

File: src/game/client/water/w3dwater.cpp [hunk 27]
```diff
@@ -0,0 +1,2327 @@
...
+void WaterRenderObjClass::Render_Water()
+{
+    for (PolygonTrigger *i = PolygonTrigger::Get_First_Polygon_Trigger(); i; i = i->Get_Next()) {
+        if (i->Is_Water_Area()) {
+            if (i->Get_Num_Points() > 2) {
+                if (i->Is_River()) {
+                    Draw_River_Water(i);
+                } else {
+                    for (int j = 1; j < i->Get_Num_Points() - 1; j++) {
+                        ICoord3D p1 = *i->Get_Point(0);
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: wrong order for ps, starts at p4, ends at p1

File: src/game/client/water/w3dwater.cpp [hunk 28]
```diff
@@ -0,0 +1,2327 @@
...
+
+void WaterRenderObjClass::Render_Water()
+{
+    for (PolygonTrigger *i = PolygonTrigger::Get_First_Polygon_Trigger(); i; i = i->Get_Next()) {
+        if (i->Is_Water_Area()) {
+            if (i->Get_Num_Points() > 2) {
+                if (i->Is_River()) {
+                    Draw_River_Water(i);
+                } else {
+                    for (int j = 1; j < i->Get_Num_Points() - 1; j++) {
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: original increments by 2?, this whole function feels off...

File: src/game/client/water/w3dwater.cpp [hunk 29]
```diff
@@ -0,0 +1,2327 @@
...
+
+        float maxy = GameMath::Ceil(gy + m_gridChangeMaxRange);
+
+        if (m_gridCellsY < maxy) {
+            maxy = m_gridCellsY;
+        }
+
+        for (int i = miny; i < maxy; i++) {
+            for (int j = minx; j < maxx; j++) {
+                WaterMeshData *mesh = &m_meshData[(m_gridCellsX + 3) * (i + 1) + 1 + j];
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: 1 + j to j + 1

File: src/game/client/water/w3dwater.cpp [hunk 30]
```diff
@@ -0,0 +1,2327 @@
...
+                float f1 = GameMath::Sqrt((gx - j) * (gx - j) + (gy - i) * (gy - i));
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: dist

File: src/game/client/water/w3dwater.cpp [hunk 31]
```diff
@@ -0,0 +1,2327 @@
...
+                float f2 =
```
Comment (tomsons26) [inline src/game/client/water/w3dwater.cpp]: height?

File: src/game/client/water/w3dwater.cpp (hunk 28 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: Fixed.

File: src/game/client/water/w3dwater.cpp (hunk 29 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 30 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 26 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 25 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 24 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 23 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp:846]: Yes.

File: src/game/client/water/w3dwater.cpp (hunk 22 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 21 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp:794]: Its there in zh wb.

File: src/game/client/water/w3dwater.cpp (hunk 20 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 19 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 18 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 21 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: I see what I did wrong, removed it. Copy and paste error from another function that does have the Ref_Ptr_Release.

File: src/game/client/water/w3dwater.cpp (hunk 17 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 16 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 15 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 14 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: Looks that way yep.

File: src/game/client/water/w3dwater.cpp (hunk 13 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp:85]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 12 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp:74]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp:750]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp:857]: fixed.

File: src/game/client/water/w3dwater.h (hunk 5 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.h]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: Fixed.

File: src/game/client/water/w3dwater.cpp (hunk 31 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: fixed.

File: src/game/client/water/w3dwater.cpp (hunk 27 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: Fixed.

File: src/game/client/water/w3dwater.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/client/water/w3dwater.cpp]: Fixed.

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**